### PR TITLE
[WB-7256] Runtime Error when logging in grad.data

### PR DESCRIPTION
**Pull request by TOsborn**
Description
-----------

https://wandb.atlassian.net/browse/WB-7256

The PR works around an issue with `torch.histc` where a runtime error can occur when `min == max > 10^15`. It also addresses a potential unnoticed issue in which the bins created by histc may not be the ones expected and may cause the histogram to be displayed incorrectly.

When `torch.histc` receives a tensor with only one distinct value, or when it receives arguments `min` and `max`, an attempt appears to be made to separate min and max by assigning `min = min-1` and `max = max+1`. However, `histc` works with 64 bit floats, so when the values are greater than roughly `10^15`, we have `min == min - 1` and `max == max + 1`, so the separation doesn't succeed, and we end up with `RuntimeError: torch.histc: max must be larger than min`. Simply calling `torch.histc(torch.tensor([1e16]))` will reproduce this; calling the same with `1e15` will succeed without issue.

The other potential issue is that the separation `torch.histc` applies results in `histc` using a different set of bins than the index we create with our call to `torch.linspace`, which can cause the histogram to be displayed incorrectly. Both of these issues are addressed.

Testing
-------

Unit tests are included.

Release Notes
-------------

------------- BEGIN RELEASE NOTES ------------------

NO RELEASE NOTES

------------- END RELEASE NOTES --------------------

### CODE REVIEW

File: tests/integrations/test_torch.py [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
+# The machine epsilon, or smallest relative difference between any two
+# numbers, for 32 and 64 bit floats. Equivalent to torch.finfo(dtype).eps
+# in pytorch v1.0.0 and later.
+EPS32 = float.fromhex("0x1p-23")
+EPS64 = float.fromhex("0x1p-52")
+
+# The smallest "normal number" for 32 and 64 bit floats. Equivalent to
+# torch.finfo(dtype).tiny in pytorch v1.0.0 and later.
+TINY32 = float.fromhex("0x1p-126")
+TINY64 = float.fromhex("0x1p-1022")
```
Comment (adrnswanberg) [inline tests/integrations/test_torch.py:24]: Can you reference these from torch.py instead of redefining?

**Review by adrnswanberg — DISMISSED** [review]
lgtm. Ideally there would be a test that fails on the case in the bug, but ok to merge with just unit tests.

File: tests/integrations/test_torch.py [hunk 2]
```diff
@@ -245,34 +250,121 @@ def test_nested_shape():
...
-    assert torch.equal(torch_history._remove_infs_nans(test_input), expected)
+    assert torch.equal(torch_history._remove_invalid_entries(test_input), expected)
+
+
+def iter_hist_bounds(dtype: torch.dtype) -> List[Tuple[float]]:
+    if dtype == torch.float32:
+        high = wandb.wandb_torch.HIGH32
+        eps = wandb.wandb_torch.EPS32
+        tiny = wandb.wandb_torch.TINY32
+    else:  # dtype == torch.float64
```
Comment (kptkin) [inline tests/integrations/test_torch.py:306]: is this a safe assumption? Types can a lot of different values. 
Should we be more explicit here?

File: tests/integrations/test_torch.py [hunk 3]
```diff
@@ -245,34 +250,121 @@ def test_nested_shape():
...
+    torch_history = wandb.wandb_torch.TorchHistory(wandb.run.history)
+    for dtype in (torch.float32, torch.float64):
+        if dtype == torch.float32:
+            high = wandb.wandb_torch.HIGH32
+            eps = wandb.wandb_torch.EPS32
+            tiny = wandb.wandb_torch.TINY32
+        else:  # dtype == torch.float64
+            high = wandb.wandb_torch.HIGH64
+            eps = wandb.wandb_torch.EPS64
+            tiny = wandb.wandb_torch.TINY64
```
Comment (kptkin) [inline tests/integrations/test_torch.py:351]: this logic is duplicated, maybe you could combine it?

File: wandb/wandb_torch.py [hunk 4]
```diff
@@ -299,18 +316,58 @@ def _torch_hook_handle_is_valid(self, handle):
         else:
             return handle.id in d
 
-    def _no_finite_values(self, tensor: "torch.Tensor") -> bool:
+    def _is_valid_mask(self, tensor: "torch.Tensor") -> bool:
+        if tensor.dtype == torch.float32:
+            high = HIGH32
+        else:  # dtype == torch.float64
```
Comment (kptkin) [inline wandb/wandb_torch.py:322]: again is this safe to assume this as default type?

File: wandb/wandb_torch.py [hunk 5]
```diff
@@ -299,18 +316,58 @@ def _torch_hook_handle_is_valid(self, handle):
...
+            high = HIGH64
+
+        return torch.logical_and((tensor <= high), (tensor >= -high))
```
Comment (kptkin) [inline wandb/wandb_torch.py:325]: notice that torch==0.4.1 doesn't have `logical_and`

File: wandb/wandb_torch.py [hunk 6]
```diff
@@ -299,18 +316,58 @@ def _torch_hook_handle_is_valid(self, handle):
...
+            or torch.logical_not(self._is_valid_mask(tensor)).all().item()
         )
 
-    def _remove_infs_nans(self, tensor: "torch.Tensor") -> "torch.Tensor":
-        if not torch.isfinite(tensor).all():
-            tensor = tensor[torch.isfinite(tensor)]
+    def _remove_invalid_entries(self, tensor: "torch.Tensor") -> "torch.Tensor":
+        if tensor.dtype == torch.float32:
+            high = HIGH32
+        else:  # dtype == torch.float64
```
Comment (kptkin) [inline wandb/wandb_torch.py:336]: same as above, about defaulting to float64

File: wandb/wandb_torch.py (hunk 4 above)
Comment (kptkin) [inline wandb/wandb_torch.py:322]: is this a safe assumption given that types can be anything?
(maybe mathematically it works out?)

File: wandb/wandb_torch.py [hunk 7]
```diff
@@ -7,13 +7,31 @@
...
+EPS32 = float.fromhex("0x1p-23")
+EPS64 = float.fromhex("0x1p-52")
+
+# The smallest "normal number" for 32 and 64 bit floats. Equivalent to
+# torch.finfo(dtype).tiny in pytorch v1.0.0 and later.
+TINY32 = float.fromhex("0x1p-126")
+TINY64 = float.fromhex("0x1p-1022")
+
 
 def nested_shape(array_or_tuple, seen=None):
```
Comment (kptkin) [inline wandb/wandb_torch.py:36]: could you add type annotation here?

File: tests/integrations/test_torch.py (hunk 2 above)
Comment (TOsborn) [reply tests/integrations/test_torch.py:306]: I checked through all of the dtypes [here](https://pytorch.org/docs/stable/tensor_attributes.html), and torch.float32 and torch.float64 are the only two supported by torch.histc, so we'd run into different issues if weights of different dtypes reached this part of the code.

File: tests/integrations/test_torch.py [hunk 8]
```diff
@@ -245,34 +250,121 @@ def test_nested_shape():
...
+            torch.tensor([0.0, 1e31, -1e31], dtype=torch.float64),
+        ),
+        (
+            torch.tensor([0.0, 1e301, -1e301], dtype=torch.float64),
+            torch.tensor([0.0], dtype=torch.float64),
+        ),
     ],
 )
-def test_remove_infs_nans(test_input, expected, wandb_init_run):
+def test_remove_invalid_entries(test_input, expected, wandb_init_run):
```
Comment (kptkin) [inline tests/integrations/test_torch.py:295]: test

**Comment by dmitryduev** [comment]
@kptkin @adrnswanberg can this be merged?

**Comment by kptkin** [comment]
> @kptkin @adrnswanberg can this be merged?

It was a while since I looked at this one... I don't remember the state of it... (not sure if all comments were addressed)

**Pull request closed without merging**